src/include/duckdb/common/types.hpp

```diff
@@ -486,10 +496,28 @@ struct StringType {
 
 struct ListType {
 	DUCKDB_API static const LogicalType &GetChildType(const LogicalType &type);
+	// Non-constant on the logical type so we can bind it in case is a user type
+	DUCKDB_API static LogicalType &GetChildTypeBind( LogicalType &type);
+};
+
+struct UserType{
+	DUCKDB_API static const string &GetTypeName(const LogicalType &type);
 };
 
+struct EnumType{
+	DUCKDB_API static const string &GetTypeName(const LogicalType &type);
+	DUCKDB_API static int64_t GetPos(const LogicalType &type, const string& key);
+	DUCKDB_API static const vector<string> &GetValuesInsertOrder(const LogicalType &type);
+	DUCKDB_API static idx_t GetSize(const LogicalType &type);
+	DUCKDB_API static shared_ptr<ExtraTypeInfo> CreateEnumInfo(const string & enum_name, const shared_ptr<vector<string>>& ordered_data);
+	DUCKDB_API static const string& GetValue(const Value &val);
+};
+
+
 struct StructType {
 	DUCKDB_API static const child_list_t<LogicalType> &GetChildTypes(const LogicalType &type);
+	// Non-constant on the logical type so we can bind it in case is a user type
+	DUCKDB_API static  child_list_t<LogicalType> &GetChildTypesBind(LogicalType &type);
```

Comment:
Types are supposed to be immutable - instead of changing a type in-place instead create a new type (if required).

Reply:
Some code exists that does this already in the `ExchangeType` method in `expression_binder.cpp:138`